File: backend/src/main/java/com/obinnaduruaku/omnimusik/sync/SyncEpoch.java
package com.obinnaduruaku.omnimusik.sync;

import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.Id;
import jakarta.persistence.Table;

import java.time.Instant;
import java.util.UUID;

/**
 * This database's identity.
 *
 * <p>Playlist version numbers are only meaningful inside the history that produced
 * them. A client that last synced at version 7 and now sees no playlist at all reads
 * that as a deletion -- correct, so long as the absence really is a deletion. Against
 * a database that was replaced, restored from a backup, or simply restarted while
 * in-memory, the same absence means "these version numbers were never mine", and
 * honouring it deletes work nobody deleted.
 *
 * <p>So the server publishes a value that changes exactly when its history does. One
 * row, written once, read on every sync. A client that sees a value other than the
 * one it stored knows its recorded versions describe a history this server has never
 * had, and re-uploads rather than deleting.
 *
 * <p>Deliberately not a timestamp or a counter: it needs no ordering, only
 * difference, and a random value cannot accidentally collide with a meaningful one
 * after a restore.
 */
@Entity
@Table(name = "sync_epoch")
public class SyncEpoch {

    /** Fixed, because there is exactly one row and a second one would be a bug. */
    public static final long SINGLETON_ID = 1L;

    @Id
    private Long id;

    /**
     * Named {@code epoch_value} rather than {@code value}: the latter is reserved in
     * H2, and Hibernate quotes nothing by default, so the generated select is a
     * syntax error at runtime rather than a mapping error at startup.
     */
    @Column(name = "epoch_value", nullable = false, updatable = false)
    private String value;

    @Column(nullable = false, updatable = false)
    private Instant createdAt;

    protected SyncEpoch() {
        // JPA
    }

    public static SyncEpoch createNew() {
        SyncEpoch epoch = new SyncEpoch();
        epoch.id = SINGLETON_ID;
        epoch.value = UUID.randomUUID().toString();
        epoch.createdAt = Instant.now();
        return epoch;
    }

    public String getValue() {
        return value;
    }

    public Instant getCreatedAt() {
        return createdAt;
    }
}

File: OmniMusik/OmniMusik/Domain/SyncEpoch.swift
//
//  SyncEpoch.swift
//  OmniMusik
//
//  Whether this device's recorded sync versions describe the server it is talking to.
//
//  `PlaylistSync` reads "the server does not have this playlist, and I last synced it
//  at version 7" as a deletion, and deleting is the right answer -- as long as the
//  absence really is a deletion. Against a database that was replaced, restored from
//  a backup, or restarted while it was in memory, the same absence means those
//  version numbers were never that server's, and honouring it destroys playlists
//  nobody deleted.
//
//  Nothing in the version numbers themselves can distinguish those two cases, which
//  is why this exists as a separate question asked before the table is consulted.
//

import Foundation

enum SyncEpochDecision: Equatable {

    /// The recorded versions came from this server's history. Reconcile as normal.
    case trustRecordedVersions

    /// They did not, or cannot be shown to have. Forget them, so every playlist looks
    /// like one that has never been uploaded and is pushed rather than deleted.
    case discardRecordedVersions
}

enum SyncEpochRule {

    /// - Parameters:
    ///   - stored: the epoch this device recorded at its last sync, or nil if it has
    ///     never recorded one.
    ///   - server: the epoch the server reports now.
    ///   - deviceHasSyncedPlaylists: whether this device holds any playlist carrying
    ///     a version it believes the server assigned.
    static func decide(
        stored: String?,
        server: String,
        deviceHasSyncedPlaylists: Bool
    ) -> SyncEpochDecision {
        guard let stored else {
            // No recorded epoch. Either this device has never synced -- in which case
            // there are no versions to distrust and nothing is at stake -- or it
            // synced before epochs existed, and there is no way to tell whether that
            // was against this database.
            //
            // Unverifiable is treated as replaced, because the two outcomes are not
            // symmetrical. Re-uploading a playlist the server already has costs a
            // request and produces a conflict a person can resolve; deleting one it
            // does not have costs the playlist, and nothing can bring it back.
            return deviceHasSyncedPlaylists ? .discardRecordedVersions : .trustRecordedVersions
        }

        return stored == server ? .trustRecordedVersions : .discardRecordedVersions
    }
}
